services/npm/npm-rating.service.js

```diff
@@ -0,0 +1,82 @@
+'use strict'
+
+const Joi = require('joi')
+const { starRating } = require('../text-formatters')
+const { BaseJsonService } = require('..')
+
+// https://api-docs.npms.io/#api-Package-GetPackageInfo
+const responseSchema = Joi.object({
+  score: Joi.object({
+    final: Joi.number().required(),
+    detail: Joi.object({
+      maintenance: Joi.number().required(),
+      popularity: Joi.number().required(),
+      quality: Joi.number().required(),
+    }),
+  }),
+}).required()
+
+module.exports = class NpmRating extends BaseJsonService {
+  static category = 'rating'
+
+  static route = {
+    base: 'npm',
+    pattern:
+      'rating/:type(maintenance|popularity|quality)?/:scope(@.+)?/:packageName',
+  }
+
+  static examples = [
+    {
+      title: 'NPM Rating',
+      namedParams: { packageName: 'egg' },
+      staticPreview: this.render({ score: 0.9711 }),
+      keywords: ['node'],
+    },
+    {
+      title: 'NPM Rating (Popularity)',
+      namedParams: { type: 'popularity', packageName: '@vue/cli' },
+      staticPreview: this.render({ type: 'popularity', score: 0.89 }),
+      keywords: ['node'],
+    },
+    {
+      title: 'NPM Rating (Quality)',
+      namedParams: { type: 'quality', packageName: 'egg' },
+      staticPreview: this.render({ type: 'quality', score: 0.98 }),
+      keywords: ['node'],
+    },
+    {
+      title: 'NPM Rating (Maintenance)',
+      namedParams: { type: 'maintenance', packageName: 'command' },
+      staticPreview: this.render({ type: 'maintenance', score: 0.222 }),
+      keywords: ['node'],
+    },
+  ]
+
+  static defaultBadgeData = {
+    label: 'Rating',
+  }
+
+  static render({ type, score }) {
+    return {
+      label: !type ? 'Rating' : type[0].toUpperCase() + type.substring(1),
+      message: starRating(score * 5),
+      color: score > 0.3 ? 'brightgreen' : 'red',
+    }
+  }
```

Comment:
The star rating seems wrong to me for representing a 0-100 percentage, especially as the only display format. Stars are really about a representation of user supplied feedback, whereas all of these numbers are really more of a score or analysis.

I probably wouldn't try to stand in the way of having stars as an opt-in formatting option if other maintainers think that's worthwhile, but I definitely don't think it should be the default. The values seem to be a canonical candidate for percentages to me.

Reply:
The logic could also be performed a bit more clearly and expressively by utilizing local variables, and we also should not be changing the case of the label text.

Finally, I think we probably want to use some kind of color scale. Switching to a percentage based formatting will make this easier since the scale is already provided, but a custom color scale can be defined as well

Reply:
are there any existing examples for the badge style?

Reply:
upgrade, now support `msg_type=grade | percentage | star` for user to choose